### DISCUSSION TITLE
testing pancel.csv

### DISCUSSION THREAD

**Discussion by Milad4849** [BodenmillerGroup/steinbock#174] [Development]
It appears that  the output of `imc.create_panel_from_imc_panel`  should be tested against `imc_test_data_steinbock_path/panel.csv` in the test dataset folder. However the outcome of `imc.create_panel_from_imc_panel(imc_test_data_steinbock_path / "raw" / "panel.csv")` does not match the file. Am I missing something here?

**Comment by jwindhager** [comment]
Could you diff the two files to see where they don't match?

**Comment by Milad4849** [comment]
<img width="717" alt="Screenshot 2023-02-24 at 01 30 22" src="https://user-images.githubusercontent.com/70942846/221062688-c4ba0913-d335-48e6-8e94-ddaebea0d475.png">
Comparing them as dataframes.

**Comment by jwindhager** [comment]
Yes, that makes sense: the only missing column seems to be `cellpose`, which was introduced with steinbock v0.16.0. The test data has not been updated in a while, so likely requires a new release.

**Reply by Milad4849** [reply]
Apart from the column, other differences also make direct comparison not possible. Shall I run the test data through the pipeline and produce new output files for a new release of the test data?

**Comment by jwindhager** [comment]
What are those other differences? Yes, it would be good to update the TestData @nilseling

**Reply by Milad4849** [reply]
The `keep` column is `bool` in first and `int` in the second file, The datatypes of some other columns are also different. NA and NaN in first, only NaN in second (I guess a consequence of the difference in datatypes). First has `channel` as a column as well as index, I think you cannot read the second file in a way that ends up like that, have to add the channel column after reading column 0 as index.  Just minor stuff that seem to have changes since the last rerelease of test data

**Reply by jwindhager** [reply]
Except for the missing column, the files should be identical on disk. If you use `steinbock.io.read_panel` instead of `pd.read_csv`, the `DataFrame` instances should be comparable (except for the index, maybe, which is undefined/not saved and is not used by steinbock - only the row order matters).

**Reply by Milad4849** [reply]
![Screenshot 2023-02-24 at 12 55 30](https://user-images.githubusercontent.com/70942846/221173810-8763cba3-3d34-4655-ab25-dafdad909e80.png)

That fixes the column datatype issue, the 'NA' vs 'NaN' and the indexes remain. It didnt seem to me that I could specify options differently to read them without these differences.

**Reply by jwindhager** [reply]
Thats because `create_panel_from_imc_panel` interprets the `ilastik`/`deepcell`/`cellpose` columns as nullable `pd.UInt8Dtype`: https://github.com/BodenmillerGroup/steinbock/blob/66e9717db847c9680f92b48bfc09ba310ffdce41/steinbock/preprocessing/imc.py#L73-L79

but `pd.read_csv` used by `read_panel` reads optional columns as pandas default data type (non-nullable float):
https://github.com/BodenmillerGroup/steinbock/blob/66e9717db847c9680f92b48bfc09ba310ffdce41/steinbock/io.py#L74-L85

As said, on disk the files should be the same, and the index is irrelevant.

**Reply by jwindhager** [reply]
It may make sense to unify this behavior (either use non-nullable or nullable column types), but I'd be careful with touching `read_panel`, since its used a lot. Maybe rather adapt `create_panel_from_imc_panel` (i.e., don't use nullable column types anywhere in the code), but this is something I would not spend too much time on, since conversion from IMC Segmentation Pipeline panels to steinbock panels should be considered a legacy feature IMHO.

**Reply by Milad4849** [reply]
All right, it is clear now, thanks for the help!

**Reply by Milad4849** [reply]
#175 should fix it
